### PR TITLE
Create Alias Domain Object and Repo

### PR DESCRIPTION
**Pull request by talanknight**

### CODE REVIEW

File: internal/alias/target/alias.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+func allocAlias() *Alias {
+	fresh := &Alias{
+		Alias: &store.Alias{},
+	}
+	return fresh
+}
+
+// NewAlias generates a new in-memory alias.
+func NewAlias(ctx context.Context, scopeId, v string, opt ...Option) (*Alias, error) {
+	const op = "alias.NewAlias"
```
Comment (irenarindos) [inline internal/alias/target/alias.go:41]: I think the ops here and elsewhere under alias/target should use target (repository_alias.go does)
```suggestion
	const op = "target.NewAlias"
```

File: internal/alias/target/repository_alias_test.go [hunk 2]
```diff
@@ -0,0 +1,856 @@
...
+			}},
+			errContains: "no value",
+		},
+		{
+			name: "no-scope",
+			in: &Alias{Alias: &store.Alias{
+				Value: "global",
+			}},
+			errContains: "no scope",
+		},
```
Comment (irenarindos) [inline internal/alias/target/repository_alias_test.go:64]: Test for an alias with a non-empty public id?

File: internal/alias/target/testing.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"context"
+	"testing"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/stretchr/testify/require"
+)
+
+func TestAlias(t *testing.T, rw *db.Db, alias string, opt ...Option) *Alias {
```
Comment (irenarindos) [inline internal/alias/target/testing.go:14]: Can you add a testing_test.go?

File: internal/daemon/controller/handlers/scopes/scope_service.go [hunk 4]
```diff
@@ -72,6 +72,8 @@ var (
 	// TODO: get this from action registry
 	scopeCollectionTypeMapMap = map[string]map[resource.Type]action.ActionSet{
 		scope.Global.String(): {
+			// TODO: uncomment this line.
+			// resource.Alias:            aliases.CollectionActions,
```
Comment (irenarindos) [inline internal/daemon/controller/handlers/scopes/scope_service.go:76]: Should this go in another PR since there should be accompanying changes in scope_service_test.go?

File: internal/alias/target/repository_alias.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+	const op = "target.(Repository).UpdateAlias"
+	switch {
+	case a == nil:
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "nil Alias")
+	case a.Alias == nil:
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "nil embedded Alias")
+	case a.PublicId == "":
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no public id")
+	case len(fieldMask) == 0:
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.EmptyFieldMask, op, "empty field mask")
```
Comment (irenarindos) [inline internal/alias/target/repository_alias.go:104]: Check for missing version

File: internal/alias/target/alias_test.go [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+				in.UpdateTime = nil
+				in.CreateTime = nil
+				assert.Empty(t, cmp.Diff(c.want, in, protocmp.Transform()))
+			}
+		})
+	}
+}
+
+func TestDelete(t *testing.T) {
+	conn, _ := db.TestSetup(t, "postgres")
```
Comment (irenarindos) [inline internal/alias/target/alias_test.go:426]: Can you add more delete cases (delete existing, delete non-existant, etc)

File: internal/alias/target/public_ids.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+	"context"
+
+	"github.com/hashicorp/boundary/globals"
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/types/resource"
+)
+
+func init() {
+	globals.RegisterPrefixToResourceInfo(globals.TargetAliasPrefix, resource.Alias, "alias", Subtype)
```
Comment (irenarindos) [inline internal/alias/target/public_ids.go:16]: I'm wondering if there should be an internal/alias/alias.go file where "alias" is defined as a const

File: internal/alias/target/repository_alias.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+)
+
+// CreateAlias inserts a into the repository and returns a new
```
Comment (irenarindos) [inline internal/alias/target/repository_alias.go:17]:
```suggestion
// CreateAlias inserts Alias a into the repository and returns a new
```

File: internal/alias/target/repository_alias.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "nil embedded Alias")
+	case a.PublicId == "":
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no public id")
+	case len(fieldMask) == 0:
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.EmptyFieldMask, op, "empty field mask")
+	case version == 0:
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no version")
+	}
+
+	var dbMask, nullFields []string
```
Comment (irenarindos) [inline internal/alias/target/repository_alias.go:109]: Is there a reason not to use the dbw.BuildUpdatePaths pattern here?(ex: https://github.com/hashicorp/boundary/blob/main/internal/credential/static/repository_credential.go#L384-L402)

File: internal/alias/target/repository_alias.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+	a := allocAlias()
+	a.PublicId = id
+	if err := r.reader.LookupByPublicId(ctx, a); err != nil {
+		if errors.IsNotFoundError(err) {
+			return db.NoRowsAffected, nil
+		}
+		return db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("failed for %s", id)))
+	}
+	if a.ScopeId == "" {
+		return db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no project id")
```
Comment (irenarindos) [inline internal/alias/target/repository_alias.go:226]:
```suggestion
		return db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no scope id")
```

File: internal/daemon/controller/rate_limiter_test.go [hunk 11]
```diff
@@ -63,7 +63,7 @@ func Test_newRateLimiterConfig(t *testing.T) {
 			ratelimit.DefaultLimiterMaxQuotas(),
 			false,
 			&rateLimiterConfig{
-				maxSize:  296148,
+				maxSize:  308154,
```
[the remark this answers is not in the document]
Comment (tmessi) [reply internal/daemon/controller/rate_limiter_test.go:66]: We could calculate it in the test, but I hesitate to do so since it would duplicate the logic that is being tested here. We would expect this to change when new endpoints are added.

File: internal/alias/target/alias_test.go [hunk 12]
```diff
@@ -0,0 +1,472 @@
...
+		a, err := NewAlias(context.Background(), "global", "with.destination", WithDestinationId("ttcp_1234567890"))
+		require.NoError(t, err)
+		assert.NotNil(t, a)
+		assert.Equal(t, a.ScopeId, "global")
+		assert.Equal(t, a.Value, "with.destination")
+		assert.Equal(t, a.DestinationId, "ttcp_1234567890")
+	})
+}
+
+func TestCreate(t *testing.T) {
```
Comment (irenarindos) [inline internal/alias/target/alias_test.go:40]: These two tests are failing:
````
FAIL: TestCreate/unsupported_project_scope (0.00s)
2024-01-29T17:30:53.8370310Z     --- FAIL: TestCreate/unsupported_org_scope (0.00s)
``
https://github.com/hashicorp/boundary/actions/runs/7700022229/job/20982976827?pr=4275
TestUpdate also has a failure

File: internal/alias/target/alias_test.go [hunk 13]
```diff
@@ -0,0 +1,472 @@
...
+				},
+			},
+		},
+		{
+			name: "remove alias value",
+			in: &Alias{
+				Alias: &store.Alias{},
+			},
+			fieldMask:   []string{"Value"},
+			errContains: `wt_target_alias_too_short constraint failed:`,
```
Comment (irenarindos) [inline internal/alias/target/alias_test.go:234]:
```suggestion
			errContains: `wt_alias_too_short constraint failed:`,
```

File: internal/alias/target/alias_test.go [hunk 14]
```diff
@@ -0,0 +1,472 @@
...
+			name:        "unsupported project scope",
+			scope:       proj.GetPublicId(),
+			value:       "unsupported.project.scope",
+			errContains: `violates foreign key constraint "iam_scope_global_fkey"`,
+		},
+		{
+			name:        "unsupported org scope",
+			scope:       proj.GetParentId(),
+			value:       "unsupported.org.scope",
+			errContains: `violates foreign key constraint "iam_scope_global_fkey"`,
```
Comment (irenarindos) [inline internal/alias/target/alias_test.go:126]: This is the failure seen instead:
`        	Error:      	"db.Create: alias_must_be_in_global_scope constraint failed: check constraint violated: integrity violation: error #1000" does not contain "violates foreign key constraint \"iam_scope_global_fkey\""`

File: internal/alias/target/alias_test.go [hunk 15]
```diff
@@ -0,0 +1,472 @@
...
+			scope:       "global",
+			value:       "host.with.no.destination",
+			opts:        []Option{WithHostId("hst_1234567890")},
+			errContains: `destination_id_set_when_host_id_is_set constraint failed`,
+		},
+		{
+			name:        "unsupported project scope",
+			scope:       proj.GetPublicId(),
+			value:       "unsupported.project.scope",
+			errContains: `violates foreign key constraint "iam_scope_global_fkey"`,
```
Comment (irenarindos) [inline internal/alias/target/alias_test.go:120]: This error seen instead: 
`        	Error:      	"db.Create: alias_must_be_in_global_scope constraint failed: check constraint violated: integrity violation: error #1000" does not contain "violates foreign key constraint \"iam_scope_global_fkey\""`

**Review by irenarindos — CHANGES_REQUESTED** [review]

**Review by irenarindos — APPROVED** [review]
